torch/linalg/__init__.py

```diff
@@ -184,3 +184,94 @@
     >>> torch.allclose(torch.tensordot(a, x, dims=x.ndim), b, atol=1e-6)
     True
 """)
+
+
+qr = _add_docstr(_linalg.linalg_qr, r"""
+qr(input, mode='reduced', *, out=None) -> (Tensor, Tensor)
+
+Computes the QR decomposition of a matrix or a batch of matrices :attr:`input`,
+and returns a namedtuple (Q, R) of tensors such that :math:`\text{input} = Q R`
+with :math:`Q` being an orthogonal matrix or batch of orthogonal matrices and
+:math:`R` being an upper triangular matrix or batch of upper triangular matrices.
+
+Depending on the value of :attr:`mode` this function returns the reduced or
+complete QR factorization. See below for a list of valid modes.
+
+.. warning:: **Differences with** :meth:`~torch.qr`:
+
+             * ``torch.qr`` takes a boolean parameter ``some`` instead of ``mode``:
+
+               - ``some=True`` is equivalent of ``mode='reduced'``: both are the
+                 default
+
+               - ``some=False`` is equivalent of ``mode='complete'``.
+
+             **Differences with** ``numpy.linalg.qr``:
+
+             * ``mode='raw'`` is not implemented
+
+             * when ``mode='r'`` numpy returns a single array `R`, whereas
+               pytorch returns a tuple `(Q, R)`, where `Q` is an empty
+               tensor.
+
+
+.. warning::
+          Backpropagation is not supported for ``mode='r'``. Use ``mode='reduced'`` instead.
+
+          If you plan to backpropagate through QR, note that the current backward implementation
+          is only well-defined when the first :math:`\min(input.size(-1), input.size(-2))`
+          columns of :attr:`input` are linearly independent.
+          This behavior will propably change once QR supports pivoting.
+
+.. note:: precision may be lost if the magnitudes of the elements of :attr:`input`
+          are large
+
+.. note:: While it should always give you a valid decomposition, it may not
```

Comment:
"This function may produce different decompositions on different device types and different platforms."

This sentence is both simpler and more precise, since CUDA, for example, doesn't use LAPACK.

Reply:
I would be in favor of keeping LAPACK/MAGMA mention here as it makes it clear to the user that this is not our implementation choice but a quirk coming from third party libraries we are using.
And they should expect such values to change when they change these libraries even with the same version of pytorch.

Reply:
If we want to mention the mathematical library used for operators, which is a fine idea, then I would just add a sentence after the proposed sentence like: "This function uses LAPACK for CPU inputs and MAGMA for CUDA inputs." Also, aren't there multiple versions of LAPACK?

Reply:
> Also, aren't there multiple versions of LAPACK?

Ho yes there are. But I don't think we have a way to print to the user which blas/lapack lib they're using.

Reply:
I tried to summarize it this way:
```
.. note:: This function uses LAPACK for CPU inputs and MAGMA for CUDA inputs,
          and may produce different decompositions on different device types
          and different platforms, depending on the precise version of the
          underlying library.
```

Reply:
Maybe add "(valid)" before "decompositions"? This note makes it seem like a flaw in the implementation, but the actual message here to the user is that numerical values in`Q` and `R` may be wildly different, and that relying on those values is user error - only the mathematical properties of the returned matrices should be relied on.

Reply:
good idea. I implemented ralph's suggestion in commit aba9124d8e, which also updates ``torch.qr`` to use the very same sentence